code/Controllers/ModelAsController.php

```diff
@@ -32,30 +32,14 @@ class ModelAsController extends Controller implements NestedController {
 	 * Get the appropriate {@link ContentController} for handling a {@link SiteTree} object, link it to the object and
 	 * return it.
 	 *
+	 * @deprecated 5.0
 	 * @param SiteTree $sitetree
 	 * @param string $action
 	 * @return ContentController
 	 */
 	public static function controller_for(SiteTree $sitetree, $action = null) {
-		if ($sitetree->class == 'SilverStripe\\CMS\\Model\\SiteTree') {
-			$controller = "SilverStripe\\CMS\\Controllers\\ContentController";
-		} else {
-			$ancestry = ClassInfo::ancestry($sitetree->class);
-			while ($class = array_pop($ancestry)) {
-				if (class_exists($class . "_Controller")) {
-					break;
-				}
-			}
-			$controller = ($class !== null)
-				? "{$class}_Controller"
-				: "SilverStripe\\CMS\\Controllers\\ContentController";
-		}
-
-		if($action && class_exists($controller . '_' . ucfirst($action))) {
-			$controller = $controller . '_' . ucfirst($action);
-		}
-
-		return class_exists($controller) ? Injector::inst()->create($controller, $sitetree) : $sitetree;
+		Deprecation::notice('5.0', 'Use $sitetree->getAssociatedController() instead');
+		return $sitetree->getAssociatedController($action);
```

Comment:
The default behaviour (inherited from `SiteTree::getControllerName()`) would be exactly what we have currently - `Page` would return `Page_Controller`, `MyPageType` would return `MyPageType_Controller`. You could then override it in your page class with something like:

```php
public function getControllerName($action = null)
{
   return 'Mysite\Controllers\PageController';
}
```

I still need to do some testing with namespaces. If you call `->getControllerName()` on `Mysite\Model\Page`, it should return `Mysite\Model\Page_Controller` to preserve BC. I’m working on the assumption that `ClassInfo::ancestry()` returns FQCNs, but I haven’t actually tested that yet.

Tbh, the `getControllerName()` method only exists for some convenience. It just saves you having to type out `return Injector::inst()->create('Some\Namespace\Class'...` in full.